streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -27,19 +49,18 @@ object ZStream {
       for {
         doneRef   <- Ref.make(false).toManaged_
         finalizer <- ZManaged.finalizerRef[R](_ => UIO.unit)
-        pull = (_: Any) =>
-          ZIO.uninterruptibleMask { restore =>
-            doneRef.get.flatMap { done =>
-              if (done) IO.fail(Left(()))
-              else
-                (for {
-                  _           <- doneRef.set(true)
-                  reservation <- managed.reserve
-                  _           <- finalizer.set(reservation.release)
-                  a           <- restore(reservation.acquire)
-                } yield a).mapError(Right(_))
-            }
+        pull = ZIO.uninterruptibleMask { restore =>
```

Comment:
The implementation of `fromEffect` in terms of `managed` is too complicated IMO. The `zio` effect has no finalizers.

I think you should add: `ZIO#once(e: E): UIO[ZIO[R, E, A]]`, which returns a ZIO effect that will execute a single time, and if you call it again, you get the specified error. You could do the same for `ZManaged`.

Then you can implement these two methods simply.